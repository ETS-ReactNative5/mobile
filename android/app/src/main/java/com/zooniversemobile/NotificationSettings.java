package com.zooniversemobile;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.google.firebase.messaging.FirebaseMessaging;

/**
 * Created by rschaaf on 11/4/16.
 */


public class NotificationSettings extends ReactContextBaseJavaModule {

    public NotificationSettings(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "NotificationSettings";
    }

    @ReactMethod
    public void setInterestSubscription(String interest, Boolean subscribed, Promise promise) {
        FirebaseMessaging messagingInstance = FirebaseMessaging.getInstance();

        if (subscribed) {
            messagingInstance.subscribeToTopic(interest);
        } else {
            messagingInstance.unsubscribeFromTopic(interest);
        }

        promise.resolve("Subscription update successful");
    }

}
