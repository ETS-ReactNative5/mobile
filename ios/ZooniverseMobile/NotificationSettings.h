//
//  NotificationSettings.h
//  ZooniverseMobile
//
//  Created by Robin Schaaf on 11/3/16.
//  Copyright Â© 2016 Zooniverse. All rights reserved.
//

#import "RCTBridgeModule.h"
#import "AppDelegate.h"

@interface NotificationSettings : NSObject <RCTBridgeModule>

@end
