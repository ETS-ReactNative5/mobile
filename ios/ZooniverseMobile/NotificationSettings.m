//
//  NotificationSettings.m
//  ZooniverseMobile
//
//  Created by Robin Schaaf on 11/3/16.
//  Copyright Â© 2016 Zooniverse. All rights reserved.
//

#import "NotificationSettings.h"
#import "RCTLog.h"

@implementation NotificationSettings

RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(setInterestSubscription:(NSString *)interest
                  subscribed:(BOOL)subscribed
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  if (subscribed) {
    [[FIRMessaging messaging] subscribeToTopic:interest];
  } else {
    [[FIRMessaging messaging] unsubscribeFromTopic:interest];
  }
  
  resolve(@"Interest subscription set");
}


@end
