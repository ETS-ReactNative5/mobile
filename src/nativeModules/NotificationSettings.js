import { NativeModules } from 'react-native'

const { NotificationSettings } = NativeModules

export const updateSubscriptionOfTopic = (subscribed, topic) => {
    const subscriptionFunction = subscribed ? subscribeToTopic : unsubscribeFromTopic
    return subscriptionFunction(topic)
}
/**
 * 
 * @param {The topic string} topic 
 * @returns A promise that resolved on completion
 */
const subscribeToTopic = (topic) => {
    return NotificationSettings.setInterestSubscription(topic, true)
}

/**
 * 
 * @param {The topic string} topic 
 * @returns A promise that resolves on completion
 */
const unsubscribeFromTopic = (topic) => {
    return NotificationSettings.setInterestSubscription(topic, false)
}