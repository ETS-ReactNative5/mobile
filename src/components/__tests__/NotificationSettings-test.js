import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { NotificationSettings } from '../NotificationSettings'

jest.mock('Switch', () => 'Switch');

const notifications = {
  'general': true
}

it('renders', () => {
  const tree = renderer.create(
    <NotificationSettings notifications={notifications} checkPushPermissions={jest.fn} pushEnabled={true} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders message if push is disabled', () => {
  const tree = renderer.create(
    <NotificationSettings notifications={{}} checkPushPermissions={jest.fn} pushEnabled={false} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
