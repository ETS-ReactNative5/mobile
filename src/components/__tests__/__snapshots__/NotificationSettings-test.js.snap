exports[`test renders 1`] = `
<View
  style={undefined}>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        undefined,
        undefined,
      ]
    } />
  <ScrollView>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          undefined,
        ]
      }>
      Zooniverse would like to occassionally send you updates about new projects or projects needing help.
    </Text>
    <View
      style={undefined}>
      <Switch
        onTintColor="rgba(0, 151, 157, 1)"
        onValueChange={[Function]}
        style={undefined}
        value={true} />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            undefined,
          ]
        }>
        General Zooniverse notifications
      </Text>
    </View>
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            undefined,
            undefined,
          ]
        }>
        Project-specific Notifications
      </Text>
    </View>
  </ScrollView>
</View>
`;

exports[`test renders message if push is disabled 1`] = `
<View
  style={undefined}>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        undefined,
        undefined,
      ]
    } />
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          undefined,
        ]
      }>
      Push notifications are not enabled on your device!  Please go to Settings > Notifications > Zooniverse to allow them.
    </Text>
  </View>
</View>
`;
